src/store.rs

```diff
@@ -0,0 +1,196 @@
+            .map(|path| self.root.join(path))
+    }
+
+    pub fn store_from_workspace(&mut self, job: &Job, workspace: Workspace) -> Result<()> {
+        let mut hasher = blake3::Hasher::new();
+        let temp = tempfile::Builder::new()
+            .prefix(&format!("rbt-job-{}", job.id))
+            .tempdir()
+            .context("couldn't create temporary directory for hashing")?;
+
+        let mut output_dirs: HashSet<PathBuf> = HashSet::new();
+
+        for output in job.outputs.iter().sorted() {
+            let source = PathBuf::from(output.as_str());
+
+            ///////////////////////////////////////////////////
+            // Step 1: Validate that the path we got is safe //
+            ///////////////////////////////////////////////////
+            for component in source.components() {
+                match component {
+                    Component::Prefix(_) | Component::RootDir =>  anyhow::bail!(
+                        "Got `{}` as an output, but absolute paths are not allowed as outputs. Remove the absolute prefix to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::ParentDir => anyhow::bail!(
+                        "Got `{}` as an output, but relative paths containing `..` are not allowed as inputs. Remove the `..` to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::CurDir | Component::Normal(_) => (),
+                };
+            }
+
+            ////////////////////////////////////////////////////////////////
+            // Step 2: add the file content to the content-addressed hash //
+            ////////////////////////////////////////////////////////////////
+            hasher.update(output.as_bytes());
+
+            let mut file = File::open(&workspace.join(output.as_str())).with_context(|| {
+                format!(
+                    "couldn't open `{}` for hashing. Did the build produce it?",
+                    output
+                )
+            })?;
+
+            // TODO: docs for Blake3 say that a 16 KiB buffer is the most
+            // efficient (for SIMD reasons), but `std::io::copy` uses an 8KiB
+            // buffer. Gonna have to do this by hand at some point to take
+            // advantage of the algorithm's designed speed.
+            std::io::copy(&mut file, &mut hasher).with_context(|| {
+                format!("could not read `{}` to calculate hash", source.display())
+            })?;
+
+            ///////////////////////////////////////////////
+            // Step 3: make sure any parent paths exist  //
+            ///////////////////////////////////////////////
+            let mut ancestors: Vec<&Path> = source.ancestors().skip(1).collect();
+            ancestors.pop(); // we've made sure this is relative, so the first item is ""
+            ancestors.reverse(); // go `[a, a/b, a/b/c]` instead of `[a/b/c, a/b, a]`
```

Comment:
In that case we'd need to pop `""` from the front (see my other comment) and would need to use `remove()` with an `O(n)` runtime, or use a `VecDeque`. `reverse()` happens in place. The docs don't specify a runtime, but I'd imagine it's like half the length since we'd be doing n/2 swaps.

But in the bigger picture, I don't expect these will be more than a handful of directories deep. There's no need for extremely deep nesting since output directories do not clash with each other in the store. The `n` here would small enough to not matter, unless I've badly misunderstood something about how iterators operate!

Reply:
Ah, so I guess you could do `source.ancestors().skip(1).rev().skip(1)` and avoid the collecting all together.

But that is also harder to understand, so I totally get it if you prefer this setup.

Reply:
I'm not sure either of those is particularly intuitive. I should probably add a comment on this similar to the one I left above!

Reply:
BTW, turns out we can't do this: `rev` is only implemented for double-ended iterators, which `ancestors` is not. ☹️